lib/logstash/inputs/tcp.rb

```diff
@@ -153,8 +159,31 @@ def server_connection_thread(output_queue, socket)
 
   def handle_socket(socket, client_address, client_port, output_queue, codec)
     peer = "#{client_address}:#{client_port}"
+    first_read = true
     while !stop?
-      codec.decode(read(socket)) do |event|
+      tbuf = read(socket)
```

Comment:
It is possible that read() returns a payload that has too-few bytes in it to form a complete header.

However, given the following part of the specification, I am willing to accept that this is unlikely my experience with modern networks:

> Both formats are designed to fit in the smallest TCP segment that any TCP/IP
host is required to support (576 - 40 = 536 bytes). This ensures that the whole
header will always be delivered at once when the socket buffers are still empty
at the beginning of a connection. The sender must always ensure that the header
is sent at once, so that the transport layer maintains atomicity along the path
to the receiver. The receiver may be tolerant to partial headers or may simply
drop the connection when receiving a partial header. Recommendation is to be
tolerant, but implementation constraints may not always easily permit this